File: app/src/main/java/no/joharei/flixr/api/models/Envelope.java
package no.joharei.flixr.api.models;

import com.google.gson.annotations.SerializedName;

import java.util.Locale;

public class Envelope<T> {
    private String stat;
    private int code;
    private String message;
    @SerializedName(value = "data", alternate = {"photosets", "photoset", "contacts", "user"})
    private T data;

    public String getStat() {
        return stat;
    }

    public T getData() {
        return data;
    }

    public int getCode() {
        return code;
    }

    public String getMessage() {
        return message;
    }

    public static class FailureException extends RuntimeException {
        final String stat;
        final int code;
        final String message;

        public FailureException(String stat, int code, String message) {
            super(String.format(Locale.getDefault(), "Stat: %s, code: %d, message: %s", stat, code, message));
            this.stat = stat;
            this.code = code;
            this.message = message;
        }
    }
}

File: app/src/main/java/no/joharei/flixr/api/models/Envelope.kt
package no.joharei.flixr.api.models

import com.google.gson.annotations.SerializedName
import java.util.*

data class Envelope<out T>(
        val stat: String,
        val code: Int,
        val message: String,
        @SerializedName(value = "data",
                alternate = arrayOf("photosets", "photoset", "contacts", "user"))
        val data: T
) {
    class FailureException(stat: String, code: Int, msg: String) : RuntimeException(
            String.format(Locale.getDefault(), "Stat: %s, code: %d, message: %s", stat, code, msg)
    )
}
